lib/utils/Database.h

```diff
@@ -277,9 +282,134 @@ class Database
      */
     enum formats {
         HDF5,
-        CSV
+        CSV,
+        HDF5_MPIO
     };
 
+    /**
+     * @brief Creates a new database file with the supplied name, with parallel I/O support.
+     *        Supported only for HDF5 format.
+     *        For HDFDatabase, the function is equivalent to create,
+     *        only extending the file name with 6 digits indicating processor rank.
+     *        For HDFDatabaseMPIO, the file is created with MPI I/O file access property.
+     *
+     * @param[in] file_name Base Name of CSV database file to create.
+     * @param[in] comm      MPI communicator to get the rank.
+     *
+     * @return True if file create was successful.
+     */
+    virtual
+    bool
+    create_parallel(
+        const std::string& file_name,
+        const MPI_Comm comm)
+    {
+        CAROM_ERROR("Abstract method Database::create_parallel!\n");
+        return false;
+    }
+
+    /**
+     * @brief Opens an existing database file with the supplied name, with parallel I/O support.
+     *        Supported only for HDF5 format.
+     *        For HDFDatabase, the function is equivalent to open,
+     *        only extending the file name with 6 digits indicating processor rank.
+     *        For HDFDatabaseMPIO, the file is opened with MPI I/O file access property.
+     *
+     * @param[in] file_name Name of existing CSV database file to open.
+     * @param[in] type      Read/write type ("r"/"wr")
+     * @param[in] comm      MPI communicator to get the rank.
+     *
+     * @return True if file open was successful.
+     */
+    virtual
+    bool
+    open_parallel(
+        const std::string& file_name,
+        const std::string& type,
+        const MPI_Comm comm)
+    {
+        CAROM_ERROR("Abstract method Database::open_parallel!\n");
+        return false;
+    }
+
+    /**
+     * @brief Writes a distributed array of doubles associated with the supplied key to
+     *        the currently open database file.
+     *        Supported only for HDF5 format.
+     *        For HDFDatabase, the function is equivalent to putDoubleArray,
+     *        writing the local array per each process.
+     *        For HDFDatabaseMPIO, the global array is written into a single file via MPI I/O.
+     *
+     * @param[in] key The key associated with the array of values to be
+     *                written.
+     * @param[in] data The array of double values to be written.
+     * @param[in] nelements The number of doubles in the array.
+     */
+    virtual
+    void
+    putDoubleArray_parallel(
```

Comment:
Would the interface be simpler if we just use the function name `putDoubleArray`, with an optional input for the type?

Reply:
I guess we could add an optional input `bool distributed=false`. This change might impact the existing applications again, but I agree this would be simpler in the long run.